### PR TITLE
Allow writes and read to not block SpanSinkImpl and only synchronize flush calls

### PR DESCRIPTION
**Pull request by bidetofevil**
## Goal

First step to persisting active spans is to sort out the concurrency story and ensuring that we don't lose data if the app crashes in a state where recorded spans are only in memory.

This change stores the flushed spans in a separate list that we can serialize before clearing the completedSpans queue.

## Testing

Added test to test race condition caused by concurrent access.

### CODE REVIEW

**Comment by bidetofevil** [comment]
* **#547** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/547?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bidetofevil and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by lucaslabari — APPROVED** [review]
LGTM

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt [hunk 1]
```diff
@@ -19,16 +21,17 @@ internal class SpanSinkImpl : SpanSink {
     }
 
     override fun completedSpans(): List<EmbraceSpanData> {
-        synchronized(completedSpans) {
-            return completedSpans.toList()
-        }
+        val spansToReturn = completedSpans.size
+        return completedSpans.take(spansToReturn)
```
Comment (lucaslabari) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt:25]: Why is this needed? Instead of just returning `completedSpans.toList()`

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt [hunk 2]
```diff
@@ -19,16 +21,17 @@ internal class SpanSinkImpl : SpanSink {
...
 
     override fun flushSpans(): List<EmbraceSpanData> {
-        synchronized(completedSpans) {
-            val flushedSpans = completedSpans.toList()
-            completedSpans.clear()
-            return flushedSpans
+        synchronized(spansToFlush) {
+            spansToFlush.set(completedSpans())
+            repeat(spansToFlush.get().size) {
+                completedSpans.removeAll(spansToFlush.get().toSet())
```
Comment (lucaslabari) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt:32]: Is this correct? removing all as many times as spans in the list?

**Review by lucaslabari — APPROVED** [review]
Left a couple of comments

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt (hunk 1 above)
Comment (bidetofevil) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt:25]: Because we're not synchronizing this and and write, we can have the completedSpans queue being modified while we are taking a snapshot. Just to be predictable especially for the flush use case, we are doing this so we aren't trying to flush events that were added while the flush was taking place.

Once the flush is triggered, everything that comes after should not be part of what's being flushed.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt (hunk 2 above)
Comment (bidetofevil) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanSinkImpl.kt:32]: hahaha this is ABSOLUTELY not correct - I will change.

I had a previous implementation that polled N times to remove and add to the flushed spans, but that leaves the the object in a weird state if it crashes in the middle. Doing this atomically works much better even if I have to convert this to a set first.
